src/App.tsx

```diff
@@ -4,31 +4,37 @@ import "./App.css";
 import logo from "./logo.svg";
 
 const App: React.FC = () => {
-  const vals =
-        [null, 3   , 1   , 6   , 7   , null, 4   , null, 9   ,
-         null, null, null, 8   , 3   , null, null, null, null,
-         8   , 2   , null, null, null, null, null, 1   , null,
-         null, 7   , 4   , null, null, 8   , 1   , 6   , null,
-         null, 8   , null, null, 6   , null, null, null, 4   ,
-         9   , null, 2   , null, null, null, null, 7   , 3   ,
-         4   , 9   , null, null, 5   , 7   , 2   , 3   , null,
-         2   , null, null, null, 9   , null, 5   , null, 7   ,
-         7   , null, 3   , 2   , null, null, 6   , null, 1   ] as Array<SquareValue | null>;
-  return (
-    <div className="App">
-      <header className="App-header">
-        <Board permanentValues={vals} />
-      </header>
-    </div>
-  );
+    const vals =
+        [null, 3, 1, 6, 7, null, 4, null, 9,
+            null, null, null, 8, 3, null, null, null, null,
+            8, 2, null, null, null, null, null, 1, null,
+            null, 7, 4, null, null, 8, 1, 6, null,
+            null, 8, null, null, 6, null, null, null, 4,
+            9, null, 2, null, null, null, null, 7, 3,
+            4, 9, null, null, 5, 7, 2, 3, null,
+            2, null, null, null, 9, null, 5, null, 7,
+            7, null, 3, 2, null, null, 6, null, 1] as Array<SquareValue | null>;
+    return (
+        <div className="App">
+            <header className="App-header">
+                <Board permanentValues={vals}/>
+            </header>
+        </div>
+    );
 };
 
 type SquareAddress = number;
 
+type SquareValue = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 ;
+
+type NumberMode = "normal" | "corner";
+
 interface BoardState {
     highlights: Set<SquareAddress>;
     contradicts: Set<SquareAddress>;
     values: Array<SquareValue | null>;
```

Comment:
values should be a map from SquareAddress to `SquareValue | null`

Reply:
it is. just in an array format. the index of each SquareValue is their corresponding SquareAddress

Reply:
I would suggest changing it to `Map<SquareAddress, SquareValue>` as it is more explicit about what you are trying to do compared to `Array<SquareValue | null>`.